src/test/scala/chiselTests/DPISpec.scala

```diff
@@ -0,0 +1,181 @@
+package chiselTests.simulator
+
+import chisel3._
+import chisel3.experimental.ExtModule
+import chisel3.simulator._
+import chisel3.util.circt.dpi._
+import circt.stage.ChiselStage
+import chisel3.util.{HasExtModuleInline, HasExtModulePath, HasExtModuleResource}
+import org.scalatest.funspec.AnyFunSpec
+import org.scalatest.matchers.must.Matchers
+import org.scalatest.matchers.should.Matchers.convertToAnyShouldWrapper
+import chisel3.util.HasBlackBoxInline
+import svsim._
+
+private object EmitDPIImplementation {
+  def apply() = {
+    val dpiImpl = s"""
+                     |#include <stdint.h>
+                     |#include <iostream>
+                     |
+                     |extern "C" void hello()
+                     |{
+                     |    std::cout << "hello from c++\\n";
+                     |}
+                     |
+                     |extern "C" void add(int lhs, int rhs, int* result)
+                     |{
+                     |    *result = lhs + rhs;
+                     |}
+  """.stripMargin
+
+    class DummyDPI extends BlackBox with HasBlackBoxInline {
+      val io = IO(new Bundle {})
+      setInline("dpi.cc", dpiImpl)
+      setInline(s"$desiredName.sv", s"module $desiredName(); endmodule")
+    }
+    val dummy = Module(new DummyDPI)
+
+  }
+}
+
+class DPIIntrinsicTest extends Module {
+  val io = IO(new Bundle {
+    val a = Input(UInt(32.W))
+    val b = Input(UInt(32.W))
+    val add_clocked_result = Output(UInt(32.W))
+    val add_unclocked_result = Output(UInt(32.W))
+  })
+
+  EmitDPIImplementation()
+
+  // Void function
+  RawClockedVoidFunctionCall("hello")(clock, true.B)
+
+  // Stateless function with result
+  val result_clocked =
+    RawClockedNonVoidFunctionCall("add", UInt(32.W), Some(Seq("lhs", "rhs")), Some("result"))(clock, true.B, io.a, io.b)
+  val result_unclocked =
+    RawUnclockedNonVoidFunctionCall("add", UInt(32.W), Some(Seq("lhs", "rhs")), Some("result"))(true.B, io.a, io.b)
+
+  io.add_clocked_result := result_clocked
+  io.add_unclocked_result := result_unclocked
+}
+
+object Hello extends DPIClockedVoidFunctionImport {
+  val functionName = "hello"
+  final def apply() = super.call()
+}
+
+object AddClocked extends DPINonVoidFunctionImport[UInt] {
+  override val functionName = "add"
+  val ret = UInt(32.W)
```

Comment:
I'm of two minds of this API here. It's common in Chisel for user APIs to be inheritance based, but I don't think it's generally the best option. I'm not sure if I have a better idea though and I don't want to sandbag so maybe we just go with it and change it later if we need to. Any thoughts @seldridge?

Reply:
There's some precedence for this kind of API for things like layers where you want to "declare" something and then refer to it multiple times. Using the object here achieves that (though I mentioned to @uenoku to self-type this with `this: Singleton =>` to mandate an object). Also, this could be slightly cleaner as an `abstract class` in order to avoid the copious overrides.

Let's try this and see where it bottoms out.

Reply:
Sounds good to me, thanks @seldridge!